### DISCUSSION TITLE
The LMI problem successfully solved but their constraints don't satisfy

### DISCUSSION THREAD

**Discussion by MostaFaramin** [yalmip/YALMIP#1015] [General discussion] [answered]
Hi Johan, 

I have a bunch of LMI constraints and I wrote codes as follows and solved them by the mosek toolbox. The results show that the problem is 'Successfully solved (MOSEK)', but when I check some of the achieved results, they don't satisfy the constraints. For example, alpha should be negative at the end, but it isn't or Eps4 should be positive but it isn't. If the problem isn't solvable, why the result would be 'Successfully solved (MOSEK)'. Or on the other hand, if everything is okay, why the constraints don't satisfy?
Thank you in advance for your help.

```
clear
clc
close all

A = [0 1 0;0 -1 1;0 -10 -5];
B = [0;0;1];
C = [1 0 0];
[np,mp] = size(B);
rp = size(C,1);

if rank(ctrb(A,B)) ~= np
    error('System is not Controllable');
end
if rank(obsv(A,C)) ~= np
    error('System is not Observable');
end

LL = lqr(A',C',1*eye(np),eye(rp))';
K1 = lqr(A,B,10*eye(np),eye(mp));
AL = A - LL*C;
Ar = A - B*K1;
m = 0;
M = m*eye(np);

% ------ LMI ----------------
P1 = sdpvar(np);
Q1 = sdpvar(mp,np);
Q2 = sdpvar(np);
P3 = sdpvar(np);
L1 = sdpvar(np,rp);
L3 = sdpvar(mp,rp);
Eps1 = sdpvar(1);
Eps2 = sdpvar(1);
Eps3 = 1e-0;
Eps4 = sdpvar(1);
alpha = sdpvar(1);
LMI11 = AL'*P1 + P1*AL + Eps1*M;
LMI12 = P1*B - C'*L3';
LMI13 = C'*L1';
LMI23 = Q1;
LMI33 = Q2*Ar' + Ar*Q2;
LMI44 = Ar'*P3 + P3*Ar;
LMI34 = P3*B*B';

LT = [LMI11     LMI12         LMI13      zeros(np)     P1           P1           P1           P1            zeros(np);...
      LMI12'    Eps4*eye(mp)  LMI23      zeros(mp,np)  zeros(mp,np) zeros(mp,np) zeros(mp,np) zeros(mp,np)  zeros(mp,np);...
      LMI13'    LMI23'        LMI33      LMI34         zeros(np)    zeros(np)    zeros(np)    zeros(np)     Q2*sqrt(m);...
      zeros(np) zeros(np,mp)  LMI34'     LMI44         zeros(np)    zeros(np)    zeros(np)    zeros(np)     zeros(np);...
      P1        zeros(np,mp)  zeros(np)  zeros(np)    -eye(np)      zeros(np)    zeros(np)    zeros(np)     zeros(np);...
      P1        zeros(np,mp)  zeros(np)  zeros(np)     zeros(np)   -Eps1*eye(np) zeros(np)    zeros(np)     zeros(np);...
      P1        zeros(np,mp)  zeros(np)  zeros(np)     zeros(np)    zeros(np)   -Eps2*eye(np) zeros(np)     zeros(np);...
      P1        zeros(np,mp)  zeros(np)  zeros(np)     zeros(np)    zeros(np)    zeros(np)   -Eps3*eye(np)  zeros(np);...
      zeros(np) zeros(np,mp)  Q2*sqrt(m) zeros(np)     zeros(np)    zeros(np)    zeros(np)    zeros(np)    -inv(Eps3)*eye(np)];

LMI = [];
LMI = [LMI, P1 >= 0];
LMI = [LMI, Q2 >= 0];
LMI = [LMI, P3 >= 0];
LMI = [LMI, Eps1 >= 0];
LMI = [LMI, Eps2 >= 0];
LMI = [LMI, Eps4 >= 0];
LMI = [LMI, alpha <= 0];
LMI = [LMI, LT <= alpha];
optimize(LMI,alpha)
P1 = double(P1);
Q2 = double(Q2);
P2 = inv(Q2);
P3 = double(P3);
Q1 = double(Q1);
L1 = double(L1);
L2 = inv(P1)*C';
L3 = double(L3);
L4 = Q1*P2;
Eps1 = double(Eps1);
Eps2 = double(Eps2);
Eps4 = double(Eps4);
double(alpha)
```

**Comment by johanlofberg** [comment, marked as answer]
https://yalmip.github.io/faq/solutionviolated/

btw, it's value not double

**Reply by MostaFaramin** [reply]
I changed the code as you said, but the results didn't change. How we can choose the right smallconstant?

clear
clc
close all

A = [0 1 0;0 -1 1;0 -10 -5];
B = [0;0;1];
C = [1 0 0];
[np,mp] = size(B);
rp = size(C,1);

LL = lqr(A',C',1*eye(np),eye(rp))';
K1 = lqr(A,B,10*eye(np),eye(mp));
AL = A - LL*C;
Ar = A - B*K1;
m = 0;
M = m*eye(np);

% ------ LMI ----------------
P1 = sdpvar(np);
Q1 = sdpvar(mp,np);
Q2 = sdpvar(np);
P3 = sdpvar(np);
L1 = sdpvar(np,rp);
L3 = sdpvar(mp,rp);
Eps1 = sdpvar(1);
Eps2 = sdpvar(1);
Eps3 = 1e-0;
Eps4 = sdpvar(1);
alpha = sdpvar(1);
LMI11 = AL'*P1 + P1*AL + Eps1*M;
LMI12 = P1*B - C'*L3';
LMI13 = C'*L1';
LMI23 = Q1;
LMI33 = Q2*Ar' + Ar*Q2;
LMI44 = Ar'*P3 + P3*Ar;
LMI34 = P3*B*B';

Aux1 = sdpvar(mp,np);
Aux2 = zeros(np);
LMI24 = Aux1;  
LMI14 = Aux2;  

LT = [LMI11     LMI12         LMI13      Aux2       P1           P1           P1           P1            zeros(np);...
      LMI12'    Eps4*eye(mp)  LMI23      Aux1          zeros(mp,np) zeros(mp,np) zeros(mp,np) zeros(mp,np)  zeros(mp,np);...
      LMI13'    LMI23'        LMI33      LMI34         zeros(np)    zeros(np)    zeros(np)    zeros(np)     Q2*sqrt(m);...
      Aux2'     Aux1'         LMI34'     LMI44         zeros(np)    zeros(np)    zeros(np)    zeros(np)     zeros(np);...
      P1        zeros(np,mp)  zeros(np)  zeros(np)    -eye(np)      zeros(np)    zeros(np)    zeros(np)     zeros(np);...
      P1        zeros(np,mp)  zeros(np)  zeros(np)     zeros(np)   -Eps1*eye(np) zeros(np)    zeros(np)     zeros(np);...
      P1        zeros(np,mp)  zeros(np)  zeros(np)     zeros(np)    zeros(np)   -Eps2*eye(np) zeros(np)     zeros(np);...
      P1        zeros(np,mp)  zeros(np)  zeros(np)     zeros(np)    zeros(np)    zeros(np)   -Eps3*eye(np)  zeros(np);...
      zeros(np) zeros(np,mp)  Q2*sqrt(m) zeros(np)     zeros(np)    zeros(np)    zeros(np)    zeros(np)    -inv(Eps3)*eye(np)];

smallconstant = 1e-10;
LMI = [];
LMI = [LMI, P1 >= smallconstant*eye(np)];
LMI = [LMI, Q2 >= smallconstant*eye(np)];
LMI = [LMI, P3 >= smallconstant*eye(np)];
LMI = [LMI, Eps1 >= smallconstant];
LMI = [LMI, Eps2 >= smallconstant];
LMI = [LMI, Eps4 >= smallconstant];
LMI = [LMI, alpha <= -smallconstant];
LMI = [LMI, LT <= alpha*eye(size(LT,1))];
optimize(LMI,alpha)
P1 = double(P1);
Q2 = double(Q2);
P2 = inv(Q2);
P3 = double(P3);
Q1 = double(Q1);
L1 = double(L1);
L2 = inv(P1)*C';
L3 = double(L3);
L4 = Q1*P2;
Eps1 = value(Eps1);
Eps2 = value(Eps2);
Eps4 = value(Eps4);
value(alpha)

**Reply by johanlofberg** [reply]
there is no right value. 10^-10 typically drowns in the general tolerances of most solvers

**Reply by johanlofberg** [reply]
and put your code in code blocks, it gets mangled in the display otherwise

**Reply by MostaFaramin** [reply]
Sorry for the many questions. Is it reasonable to select a specific small-constant for each constraint separately?

**Reply by johanlofberg** [reply]
of course (but most often you would not bother with this at all. is it really that critical for the application at hand i.e. is a feasibility error 10^-8 really a problem)
